tracer/src/Datadog.Trace/DiagnosticListeners/AspNetCoreDiagnosticObserver.cs

```diff
@@ -765,12 +765,15 @@ private void OnHostingHttpRequestInStop(object arg)
 
             if (scope != null)
             {
-                // if we had an unhandled exception, the status code is already updated
-                if (!scope.Span.Error && arg.TryDuckCast<HttpRequestInStopStruct>(out var httpRequest))
+                var span = scope.Span;
+
+                // If we had an unhandled exception, the status code will already be updated correctly,
+                // but if the span was manually marked as an error, we still need to record the status code
+                if (!span.HasHttpStatusCode() && arg.TryDuckCast<HttpRequestInStopStruct>(out var httpRequest))
                 {
                     HttpContext httpContext = httpRequest.HttpContext;
-                    scope.Span.SetHttpStatusCode(httpContext.Response.StatusCode, isServer: true);
-                    scope.Span.SetHeaderTags(new HeadersCollectionAdapter(httpContext.Response.Headers), tracer.Settings.HeaderTags, defaultTagPrefix: SpanContextPropagator.HttpResponseHeadersTagPrefix);
+                    span.SetHttpStatusCode(httpContext.Response.StatusCode, isServer: true);
+                    span.SetHeaderTags(new HeadersCollectionAdapter(httpContext.Response.Headers), tracer.Settings.HeaderTags, defaultTagPrefix: SpanContextPropagator.HttpResponseHeadersTagPrefix);
```

Comment:
There is a slight change in behavior here. Before we would be adding the header tags whenever there was no error. Now you only add them if there is no status code. It means that if there is a status code and no error, we would previously set the header tags and not anymore.

Reply:
Hmmm, that's a good point. That behaviour is in `master` too 🤔 I wonder then if we should change it so that the headers are _always_ added in the stop, regardless (which would supersede #2460 too)

Reply:
https://github.com/DataDog/dd-trace-dotnet/pull/2480